File: pki.php
<?php
class PKI {
    private static $file = 'pki_keys.bin';
    private static $curve = OPENSSL_CURVE_PRIME256V1;
    private static $rsaBits = 2048;
    private static $signingPriv, $signingPub, $encryptingPriv, $encryptingPub;

    public static function init() {
        if (file_exists(self::$file)) {
            $keys = unserialize(file_get_contents(self::$file));
            self::$signingPriv = $keys['signing_priv'];
            self::$signingPub = $keys['signing_pub'];
            self::$encryptingPriv = $keys['encrypting_priv'];
            self::$encryptingPub = $keys['encrypting_pub'];
        } else {
            // Generate signing keypair (ECDSA P-256)
            $signingConfig = [
                'private_key_type' => OPENSSL_KEYTYPE_EC,
                'curve_name' => self::$curve,
            ];
            $signingRes = openssl_pkey_new($signingConfig);
            openssl_pkey_export($signingRes, $signingPrivPem);
            self::$signingPriv = $signingPrivPem;
            $signingPubDetails = openssl_pkey_get_details($signingRes);
            self::$signingPub = $signingPubDetails['key'];

            // Generate encrypting keypair (RSA 2048)
            $encryptingConfig = ['private_key_bits' => self::$rsaBits, 'private_key_type' => OPENSSL_KEYTYPE_RSA];
            $encryptingRes = openssl_pkey_new($encryptingConfig);
            openssl_pkey_export($encryptingRes, $encryptingPrivPem);
            self::$encryptingPriv = $encryptingPrivPem;
            $encryptingPubDetails = openssl_pkey_get_details($encryptingRes);
            self::$encryptingPub = $encryptingPubDetails['key'];

            // Serialize to file
            $keys = [
                'signing_priv' => self::$signingPriv,
                'signing_pub' => self::$signingPub,
                'encrypting_priv' => self::$encryptingPriv,
                'encrypting_pub' => self::$encryptingPub,
            ];
            file_put_contents(self::$file, serialize($keys));
        }
    }

    public static function sign($message) {
        self::init();
        openssl_sign($message, $signature, self::$signingPriv, OPENSSL_ALGO_SHA256);
        return base64_encode($signature);
    }

    public static function verify($message, $signatureB64) {
        self::init();
        $signature = base64_decode($signatureB64);
        $result = openssl_verify($message, $signature, self::$signingPub, OPENSSL_ALGO_SHA256);
        return $result === 1;
    }

    public static function encrypt($message) {
        self::init();
        openssl_public_encrypt($message, $ciphertext, self::$encryptingPub, OPENSSL_PKCS1_OAEP_PADDING);
        return base64_encode($ciphertext);
    }

    public static function decrypt($ciphertextB64) {
        self::init();
        $ciphertext = base64_decode($ciphertextB64);
        openssl_private_decrypt($ciphertext, $plaintext, self::$encryptingPriv, OPENSSL_PKCS1_OAEP_PADDING);
        return $plaintext;
    }
}

// Usage example:
PKI::init();
$sig = PKI::sign('Hello World');
echo "Signature: " . $sig . "\n";
echo "Verify: " . (PKI::verify('Hello World', $sig) ? 'true' : 'false') . "\n";
$enc = PKI::encrypt('Secret message');
echo "Encrypted: " . $enc . "\n";
echo "Decrypted: " . PKI::decrypt($enc) . "\n";
?>

File: pki.java
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.*;
import java.security.spec.ECGenParameterSpec;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import javax.crypto.Cipher;
import javax.crypto.EncryptedPrivateKeyInfo;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.PBEParameterSpec;

public class PKI {
    private static final String FILE = "pki_keys.bin";
    private static final String CURVE = "secp256r1";
    private static final int RSA_BITS = 2048;
    private static PrivateKey signingPriv;
    private static PublicKey signingPub;
    private static PrivateKey encryptingPriv;
    private static PublicKey encryptingPub;

    public static void init() throws Exception {
        File file = new File(FILE);
        if (file.exists()) {
            try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(file))) {
                signingPriv = (PrivateKey) ois.readObject();
                signingPub = (PublicKey) ois.readObject();
                encryptingPriv = (PrivateKey) ois.readObject();
                encryptingPub = (PublicKey) ois.readObject();
            }
        } else {
            // Generate signing keypair (ECDSA)
            KeyPairGenerator signingGen = KeyPairGenerator.getInstance("EC");
            signingGen.initialize(new ECGenParameterSpec(CURVE));
            KeyPair signingPair = signingGen.generateKeyPair();
            signingPriv = signingPair.getPrivate();
            signingPub = signingPair.getPublic();

            // Generate encrypting keypair (RSA)
            KeyPairGenerator encryptingGen = KeyPairGenerator.getInstance("RSA");
            encryptingGen.initialize(RSA_BITS);
            KeyPair encryptingPair = encryptingGen.generateKeyPair();
            encryptingPriv = encryptingPair.getPrivate();
            encryptingPub = encryptingPair.getPublic();

            // Serialize to file
            try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(file))) {
                oos.writeObject(signingPriv);
                oos.writeObject(signingPub);
                oos.writeObject(encryptingPriv);
                oos.writeObject(encryptingPub);
            }
        }
    }

    public static String sign(String message) throws Exception {
        if (signingPriv == null) init();
        Signature sig = Signature.getInstance("SHA256withECDSA");
        sig.initSign(signingPriv);
        sig.update(message.getBytes("UTF-8"));
        byte[] signature = sig.sign();
        return Base64.getEncoder().encodeToString(signature);
    }

    public static boolean verify(String message, String signatureB64) throws Exception {
        if (signingPub == null) init();
        Signature sig = Signature.getInstance("SHA256withECDSA");
        sig.initVerify(signingPub);
        sig.update(message.getBytes("UTF-8"));
        byte[] signature = Base64.getDecoder().decode(signatureB64);
        return sig.verify(signature);
    }

    public static String encrypt(String message) throws Exception {
        if (encryptingPub == null) init();
        Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
        cipher.init(Cipher.ENCRYPT_MODE, encryptingPub);
        byte[] ciphertext = cipher.doFinal(message.getBytes("UTF-8"));
        return Base64.getEncoder().encodeToString(ciphertext);
    }

    public static String decrypt(String ciphertextB64) throws Exception {
        if (encryptingPriv == null) init();
        Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
        cipher.init(Cipher.DECRYPT_MODE, encryptingPriv);
        byte[] ciphertext = Base64.getDecoder().decode(ciphertextB64);
        byte[] plaintext = cipher.doFinal(ciphertext);
        return new String(plaintext, "UTF-8");
    }

    public static void main(String[] args) throws Exception {
        init();
        String sig = sign("Hello World");
        System.out.println("Signature: " + sig);
        System.out.println("Verify: " + verify("Hello World", sig)); // true
        String enc = encrypt("Secret message");
        System.out.println("Encrypted: " + enc);
        System.out.println("Decrypted: " + decrypt(enc)); // Secret message
    }
}
